### PR TITLE
Laboratory work #2, Lilia Sabirova - 23FPL1

### PR DESCRIPTION
**Pull request by SabirovaLilia**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -18,6 +18,23 @@ def tokenize(text: str) -> list[str] | None:
...
     """
+    if not isinstance(text, str):
+        return None
+    tokenized_list = []
+    word = []
+
+    for element in text.lower():
+        if element.isalpha():
+            word.append(element)
+        else:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:29]: add `continue` to the `else` statement and remove `else`

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -33,6 +50,25 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+            isinstance(tokens, list) and
+            all(isinstance(token, str) for token in tokens)
+    ):
+        return None
+    if not (
+            isinstance(stopwords, list) and
+            all(isinstance(word, str) for word in stopwords)
+    ):
+        return None
+    if len(tokens) == 0 or len(stopwords) == 0:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:63]: try thruthy pattern: https://stackoverflow.com/questions/39983695/what-is-truthy-and-falsy-how-is-it-different-from-true-and-false

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (SabirovaLilia) [reply lab_2_retrieval_w_bm25/main.py:29]: removing ‘else’ leads to errors in the checks. why does this happen?

File: lab_2_retrieval_w_bm25/main.py (hunk 2 above)
Comment (SabirovaLilia) [reply lab_2_retrieval_w_bm25/main.py:63]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (artyomtugaryov) [reply lab_2_retrieval_w_bm25/main.py:30]: I do not know

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Hope you will get 10

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (artyomtugaryov) [reply lab_2_retrieval_w_bm25/main.py:30]: probably, I forgot about continue

**Review by artyomtugaryov — DISMISSED** [review]

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -62,6 +115,28 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
...
+            tf_dict[word] = 0.0
+        for elem in vocab:
+            if elem not in tf_dict:
+                tf_dict[elem] = 0.0
+            tf_dict[word] = document_tokens.count(word) / len(document_tokens)
+
+    if not tf_dict:
+        return None
+
+    return tf_dict
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:139]: return tf_dict or None

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -92,6 +193,26 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
...
+
+    tf_idf = {}
+    for elem, term_frequency in tf.items():
+        if elem in idf:
+            tf_idf[elem] = term_frequency * idf[elem]
+
+    if not tf_idf:
+        return None
+
+    return tf_idf
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:215]: the same

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -138,6 +294,48 @@ def rank_documents(
...
+    for i in range(len(ranged_documents) - 1):
+        for j in range(len(ranged_documents) - i - 1):
+            if ranged_documents[j][1] < ranged_documents[j + 1][1]:
+                (ranged_documents[j], ranged_documents[j + 1]) = (ranged_documents[j + 1],
+                                                                  ranged_documents[j])
+
+    if not ranged_documents:
+        return None
+
+    return ranged_documents
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:338]: the same

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -168,6 +366,44 @@ def calculate_bm25_with_cutoff(
...
+        bm25_score = (idf_document[elem] *
+                      ((frequency * (k1 + 1)) / (frequency + k1 * (
+                              1 - b + b * (doc_len / avg_doc_len)
+                      ))))
+        bm25_with_cutoff_dict[elem] = bm25_score
+
+    if not bm25_with_cutoff_dict:
+        return None
+
+    return bm25_with_cutoff_dict
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:406]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (SabirovaLilia) [reply lab_2_retrieval_w_bm25/main.py:135]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (SabirovaLilia) [reply lab_2_retrieval_w_bm25/main.py:208]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 6 above)
Comment (SabirovaLilia) [reply lab_2_retrieval_w_bm25/main.py:396]: done

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -33,6 +52,25 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+        return None
+    if not (
+            isinstance(stopwords, list) and
+            all(isinstance(word, str) for word in stopwords)
+    ):
+        return None
+    if not (len(tokens) and len(stopwords)):
+        return None
+
+    tokens_without_stopwords = []
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:68]: try to do it in one line

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (artyomtugaryov) [reply lab_2_retrieval_w_bm25/main.py:135]: really?

File: lab_2_retrieval_w_bm25/main.py [hunk 8]
```diff
@@ -77,6 +148,32 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+               all(isinstance(sub_elem, str) for sub_elem in element)
+               for element in documents):
+        return None
+
+    idf_dict = {}
+
+    for doc in documents:
+        for token in set(doc):
+            if token in vocab:
+                if not token in idf_dict:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:167]: wrong condition

File: lab_2_retrieval_w_bm25/main.py [hunk 9]
```diff
@@ -77,6 +148,32 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+                    idf_dict[token] = 1
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:168]: set 0 if token not in dict and do +1 always

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (artyomtugaryov) [reply lab_2_retrieval_w_bm25/main.py:208]: really?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**